### DISCUSSION TITLE
Possible performance issue in using plugin parameter forecasters with hierarchical data

### DISCUSSION THREAD

**Discussion by yarnabrina** [alan-turing-institute/sktime#5056] [General]
## Reproducible Example

```python
import time
import warnings

from sklearn.preprocessing import RobustScaler

from sktime.forecasting.compose import (
    ForecastingPipeline,
    ForecastX,
    TransformedTargetForecaster,
)
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.statsforecast import StatsForecastAutoARIMA
from sktime.param_est.plugin import PluginParamsForecaster
from sktime.param_est.seasonality import SeasonalityACF
from sktime.transformations.compose import TransformerPipeline
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.impute import Imputer
from sktime.utils._testing.hierarchical import _make_hierarchical

warnings.filterwarnings("ignore")

df = _make_hierarchical(
    hierarchy_levels=(2, 2),
    max_timepoints=(365 * 5),
    min_timepoints=(365 * 5),
    n_columns=5,
    all_positive=False,
    random_state=0,
    add_nan=True,
)

y = df[["c0"]]
X = df[["c1", "c2", "c3", "c4"]]

y_train, y_test, X_train, X_test = temporal_train_test_split(
    y, X=X, test_size=(365 * 2)
)

median_imputer = Imputer(method="median")
nearest_imputer = Imputer(method="nearest")

sklearn_scaler = RobustScaler()
sktime_scaler = TabularToSeriesAdaptor(RobustScaler())

periodicity_estimator = SeasonalityACF()
auto_arima_algorithm = StatsForecastAutoARIMA()
plugin_forecaster = PluginParamsForecaster(
    periodicity_estimator, auto_arima_algorithm, params={"sp": "sp"}
)

model = ForecastX(plugin_forecaster, plugin_forecaster)

X_pipeline = TransformerPipeline([nearest_imputer, sktime_scaler])
y_pipeline = TransformedTargetForecaster([median_imputer, model])

pipeline = ForecastingPipeline([X_pipeline, y_pipeline])


def get_training_time(
    training_pipeline: ForecastingPipeline, training_y, training_X
) -> float:
    tic = time.perf_counter()
    training_pipeline.clone().fit(training_y, X=training_X, fh=(30 * 2))
    toc = time.perf_counter()

    return toc - tic


print(get_training_time(pipeline, y_train, X_train))

print(
    get_training_time(
        pipeline, y_train.xs(("h0_0", "h1_0")), X_train.xs(("h0_0", "h1_0"))
    )
)
print(
    get_training_time(
        pipeline, y_train.xs(("h0_0", "h1_1")), X_train.xs(("h0_0", "h1_1"))
    )
)
print(
    get_training_time(
        pipeline, y_train.xs(("h0_1", "h1_0")), X_train.xs(("h0_1", "h1_0"))
    )
)
print(
    get_training_time(
        pipeline, y_train.xs(("h0_1", "h1_1")), X_train.xs(("h0_1", "h1_1"))
    )
)
```

## Results

```sh
❯ python plugin_test.py
47.420656906999284
31.68732754499979
1.2294460339999205
3.103604627000095
3.301523888999327

❯ python plugin_test.py
47.826508039000146
30.9390308359998
1.2870449459996962
2.941016815999319
3.6795935230002215
```

## Version

`main` branch

## Comments

This is a follow-up from a discussion I started in the discord. I was using a subset of M5 and noted that running on hierarchical was taking me >20 minutes (I didn't wait and stopped), but running individually one by one was taking just about ~1 minutes. I could not reproduce that much of difference with this random dataset, but it still shows some ~10s difference between hierarchical with 4 series and total of 4 individual series, which is close to 25%.

**Comment by yarnabrina** [comment]
FYI @fkiraly

**Comment by fkiraly** [comment]
Will look into this, what I find odd though is that the first series takes 30sec, and the other three about 2sec.

Do you have any explanation?

**Reply by fkiraly** [reply]
also, from looking at this, I would indeed believe that, at least in theory, the "all together" and "separately" ways of execution should be equivalent, logic-wise. None of the forecasters or transformers in question are genuinely hierarchical, so all broadcast.

**Reply by fkiraly** [reply]
btw, have you tried to see what happens on this branch?
https://github.com/sktime/sktime/pull/5036

**Reply by fkiraly** [reply]
spotted one opportunity for speedup, not sure whether that makes a big difference: https://github.com/sktime/sktime/pull/5057

**Comment by yarnabrina** [comment]
I do not have any explanation for what's special about first series.

I ran the scripts again on different branches for benchmarking, and here are the results. I'm not sure what happened to my laptop/environment, but the times have increased consistently. All branches still produce the difference though.

on `main` branch (at commit e58d7e991)

```sh
❯ python plugin_test.py
61.047856866000004
38.486002722999984
1.8708789490000015
3.9301593630000013
5.505913585999991
```

on `speed-trafo-boilerplate` branch (for PR #5036 at commit 9baa8dcdd)

```sh
❯ python plugin_test.py
61.574225579
37.60583952000002
1.7013179209999976
3.189458400000035
3.9899349800000437
```

on `speedup-paramfitter` branch (for PR #5057 at commit b05fb9f71)

```sh
❯ python plugin_test.py
56.35312394099998
37.036641463000024
1.7552324669999848
3.1396764499999676
3.9343035960000066
```

**Reply by fkiraly** [reply]
> I'm not sure what happened to my laptop/environment, but the times have increased consistently.

Are you perhaps on a different branch? E.g., were you initially on one of the speedup fixes?

**Reply by fkiraly** [reply]
PS: I also do not get why the speedup seems to be a proportion of total runtime. Strange and mysterious.

**Comment by fkiraly** [comment]
Here are my runtimes, by the way, with plain `ARIMA` (from `pmdarima`), on `main`:

```
2.0618726001121104
0.5104135998990387
0.5029854001477361
0.5086958000902086
0.5130260000005364
```

With `AutoARIMA`, it's run so far for 4mins and I've killed it (it was somehwere in optimization of log-likelihood)

Maybe the 1st series gives a difficult to optimize log-likelihood?

**Reply by fkiraly** [reply]
(not much difference with `speedup-trafo-boilerplate`, btw)

**Reply by fkiraly** [reply]
I observe, though, over the range of minutes, ups and downs in runtime of 50%. Perhaps background processes slowing it down?

**Reply by yarnabrina** [reply]
@fkiraly I also do get similar results with `ARIMA`, but does it use `sp` parameter? Since it takes seasonality parameters as a tuple of 4, I don't think plugin is actually happening or making any effect.

When I skip plugin part, I don't observe (at least not so far) any performance issue for any model.

**Reply by fkiraly** [reply]
that's very odd. I have no explanation currently - the `BaseParamFitter` doesn't seem to have an obvious part that would cause it, to me.

have you tried profiling? (`py-spy` etc)

**Reply by fkiraly** [reply]
also, are you able to reproduce this with `statsmodels` or only with core dependencies?

**Reply by yarnabrina** [reply]
Not in a reproducible way, but I noticed a weird pattern, which can be by chance. I tried with `NaiveForecaster` and `ExponentialSmoothing`, and for both the first time I ran the script showed time difference. Not by much, around 1 second for entire dataset and around 0.6 seconds as sum of per series. But this difference was not present from 2nd time onwards.

I wondered if `__pycache__` or something else is affecting this or not, but found nothing conclusive.

---

~I did not try profiling, actually don't know how to use `py-spy`.~

I tried `py-spy`, but I don't know how to interpret from this.

![plugin_test_pyspy_profiling](https://github.com/sktime/sktime/assets/39331844/5fc8396c-b271-49a1-ac56-48cf6c43547d)

Branch: `main`
Commit: 0094f868c
Command: `py-spy record --output plugin_test_pyspy_profiling.svg -- python plugin_test.py`
Outputs:

```sh
❯ py-spy record --output plugin_test_pyspy_profiling.svg -- python plugin_test.py
py-spy> Sampling process 100 times a second. Press Control-C to exit.

86.41467029299997
55.44373545899998
2.252269000999945
4.6028416760000255
5.50727413300001

py-spy> Stopped sampling because process exited
py-spy> Wrote flamegraph data to 'plugin_test_pyspy_profiling.svg'. Samples: 16026 Errors: 0
```
